EIPS/eip-7002.md

```diff
@@ -0,0 +1,380 @@
+        require(sent, "Failed to return excess fee payment")
+
+######################
+# Additional Helpers #
+######################
+        
+def get_exit_fee() -> int:
+    excess_exits = sload(VALIDATOR_EXIT_PRECOMPILE_ADDRESS, EXCESS_EXITS_STORAGE_SLOT)
+    return fake_exponential(
+        MIN_EXIT_FEE,
+        excess_exits,
+        EXIT_FEE_UPDATE_FRACTION
+    )
+    
+def fake_exponential(factor: int, numerator: int, denominator: int) -> int:
+    i = 1
+    output = 0
+    numerator_accum = factor * denominator
+    while numerator_accum > 0:
+        output += numerator_accum
+        numerator_accum = (numerator_accum * numerator) // (denominator * i)
+        i += 1
+    return output // denominator
+```
+
+##### Gas cost
+
+TBD
+
+Once functionality is reviewed and solidified, we'll estimate the cost of running the above computations fully in the EVM, and then potentially apply some discount due to reduced EVM overhead of being able to execute the above logic natively.
+
+#### Block structure
+
+Beginning with the `FORK_BLOCK`, the block body **MUST** be appended with a list of exit operations. RLP encoding of the extended block body structure **MUST** be computed as follows:
+
+```python
+block_body_rlp = RLP([
+    field_0,
+    ...,
+    # Latest block body field before `exits`
+    field_n,
+    
+    [exit_0, ..., exit_k],
+])
+```
+
+Beginning with the `FORK_BLOCK`, the block header **MUST** be appended with the new **`exits_root`** field. The value of this field is the trie root committing to the list of exits in the block body. **`exits_root`** field value **MUST** be computed as follows:
+
+```python
+def compute_trie_root_from_indexed_data(data):
+    trie = Trie.from([(i, obj) for i, obj in enumerate(data)])
+    return trie.root
+
+block.header.exits_root = compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+#### Block validity
+
+Beginning with the `FORK_BLOCK`, client software **MUST** extend block validity rule set with the following conditions:
+
+1. Value of **`exits_root`** block header field equals to the trie root committing to the list of exit operations contained in the block. To illustrate:
+
+```python
+def compute_trie_root_from_indexed_data(data):
+    trie = Trie.from([(i, obj) for i, obj in enumerate(data)])
+    return trie.root
+
+assert block.header.exits_root == compute_trie_root_from_indexed_data(block.body.exits)
+```
+
+2. The list of exit operations contained in the block body **MUST** be equivalent to list of exits at the head of the exit precompile's exit message queue up to the maximum of `MAX_EXITS_PER_BLOCK`, respecting the order in the queue. This validation **MUST** be run after all transactions in the current block are processed and **MUST** be run before per-block precompile storage calculations (i.e. a call to `update_exit_precompile()`) are performed. To illustrate:
+
+```python
+class ValidatorExit(object):
```

Comment:
have updated to `ValidatorExit` rather than just exit,

should the field names be update to `validatior_exits` in block/payload etc. although exits in the block is pretty clear to what is being refereed , so letting it stay